### ISSUE TITLE
Select component

### ISSUE THREAD

**Issue by Vynlar** [chakra-ui/chakra-ui#38]
@segunadebayo @tioluwani94 Really great work on everything so far!

I notice in the roadmap that you are planning to build a custom select component in Chakra and I wanted to start a discussion about how that would look.

In my experience for real-world projects, select boxes often need features that are not provided by the native HTML select such as search, multi-select, and dynamic option loading. I frequently turn to the `react-select` package because I feel it has the best API and available features. How do you see Chakra handling this problem, if at all? I see a few options:

- Chakra provides a simplistic select that doesn't have these "fancy" features and it is the responsibility of the end-developer to use a third party library if necessary
- Chakra writes its own select that does implement these features
- Chakra wraps an existing library like `react-select`

I am in favor of options 1 and/or 3 and would like to hear the community's opinions about the pros/cons of each.

Side note: This issue is also going to apply to other "complex" components such as date pickers

**Comment by segunadebayo** [comment]
Hi @Vynlar .

I'm currently trying to figure out the API for Chakra's Custom Select. I'd prefer we wrap an existing library and make the final API really simple and composable.

I also use `react-select` as well but I'm a bit reluctant to use it in Chakra due to its size. I'm currently testing `use-downshift`, a hook based library built for `downshift` and it's quite interesting and should give us the composition and functionality we need.

Here's how I envision the API for Chakra's Simple Select (pretty similar to native):

```jsx
// Simple Select
<Select isSearchable={true} value="..." defaultValue="...">
   <Option value="...">Option 1</Option>
   <Option value="...">Option 2</Option>
</Select>
```

```jsx
// Grouped Select
<Select isSearchable={true} value="..." defaultValue="...">
  <OptionGroup label="Option Group">
     <Option value="...">Option 1</Option>
    <Option value="...">Option 2</Option>
  </OptionGroup>
</Select>
```

The advantage of using `use-downshift` is that we can compose Autocomplete, Typeahead, and Multiselect out of it. Check out the docs: https://use-downshift.now.sh/docs-dropdown

Let me know what you think @Vynlar .

PS: I'm currently playing with it in CodeSandbox: https://codesandbox.io/s/chakra-playground-3tbl9

**Comment by Vynlar** [comment]
@segunadebayo `use-downshift` looks like a much better solution for Chakra than `react-select` does. It should let us create a great API that is homogenous with the existing Chakra components while still providing the features needed.

Looking forward to the first draft you come up with. Let the community know if we can help with any of these things.

**Comment by segunadebayo** [comment]
Thank you @Vynlar .

I do need help figuring out the `react-popper` issue #31  with the `<MenuButton/>`. I haven't looked at it yet and it'll be good to have someone look into it.

I'd be happy if you can jump on it.

**Comment by fernandobandeira** [comment]
I think you can't go wrong with `use-downshift` it's smaller, simpler and the flexibility it provides is the most important thing for projects like this.

However, I'd suggest that if we're going to follow the 1st route from now on, then we should be more clear about what's contained inside "fancy", putting down the minimum api requirements would help before making decisions and suggestions, you've added some requirements in there but if we combine this with the `ComboBox` and the `Tags Input` that are also in the roadmap and can be considered as variants of this component, then it might make sense to use `react-select` since it'd cover these two as well.
tl;dr: The requirements has a direct impact in deciding how we implement the core component.

Also: great work guys, we've started using this on @cheerfulgiving it's been amazing so far, hope that we can contribute to this project ❤️

**Comment by segunadebayo** [comment]
Hi @fernandobandeira ,

Thank you for using Chakra UI @cheerfulgiving.

I'll create an issue that describes the minimum API requirements for the `Select` & `ComboBox` components. Then I'll get feedback from the community before building it.

Thanks for that suggestion.

**Comment by manzoorwanijk** [comment]
> I'll create an issue that describes the minimum API requirements for the Select & ComboBox components.

@segunadebayo Please reference this issue when you do so, just to be able to follow the updates. I am eagerly waiting for Multi Select, the only feature that's stopping me from using Chakra in my two current projects.

Thank you for the library 👏

**Comment by eranimo** [comment]
Any updates on this? If work hasn't already started, I would be interested in contributing. A robust select component would really make this library feel complete.

**Comment by segunadebayo** [comment]
Hi @eranimo,

No work has started on it yet. I'll be happy to have you take a shot at the Select component. I'll open a new Issue for the API requirements of the Select component, then you can start some work once it's clear to you.

Thank you for choosing to contribute. Can't wait to see a Select component for Chakra.

**Comment by cheapernepalstudio** [comment]
Here is another good one from Uber 

https://baseweb.design/components/select/

**Comment by eranimo** [comment]
@segunadebayo Any update on the API requirements?

**Comment by segunadebayo** [comment]
Hi @eranimo @fernandobandeira, I just added the API requirements for the Single Select component.

Let me know your thoughts.

**Comment by Billmike** [comment]
@eranimo Has work begun yet on this component?

**Comment by eranimo** [comment]
@Billmike Yes, I have started. Things have been crazy at work recently though, but I'll post a PR this weekend.

**Comment by Billmike** [comment]
Awesome then!

**Comment by kermopajula** [comment]
Hey, any updates on this?

**Comment by manzoorwanijk** [comment]
It's been too long for me to wait for multiselect in Chakra UI, looks like I have to continue my struggle with AntDesign! 😞

**Comment by segunadebayo** [comment]
I've worked on the custom `Select` component but not the multiselect. It'll be released in the n ext version.

**Comment by eranimo** [comment]
Sorry, some things came up and I've been delayed getting this to a good point to PR.

**Comment by segunadebayo** [comment]
I'm closing this out. Check updates on #140

**Comment by monecchi** [comment]
@segunadebayo I totally agree with a custom approach for building a powerful Chakra-UI `Select` component! Here's one using `downshift`... It covers pretty much the basics such as array of string or objects, autocomplete etc ... But it's a start point... https://github.com/catho/quantum/tree/master/components/Dropdown
https://catho.github.io/quantum/?path=/story/forms--dropdown

**Comment by csandman** [comment]
I'm not sure what the progress on this is but I wanted to share my wrapper for react-select's multi component. I needed something like this to use with Chakra and wanted it to look the same so I spent a decent amount of time making it match visually. If you want to use it, here's the source. Feel free to modify it as needed until an official Chakra version comes out.

- Gist: https://gist.github.com/csandman/c687a9fb4275112f281ab9a5701457e4
- CodeSandbox: https://codesandbox.io/s/chakra-ui-react-select-648uv?file=/multi-select.js

I'm also curious if anyone has any recommendations to change this, I'm still very new to the inner workings of Chakra and I'm sure I didn't do everything in the best way.

**Comment by csandman** [comment]
For anyone that has had any interest in my wrapper for react-select, I finally made it into an NPM package! You can find it here: https://www.npmjs.com/package/chakra-react-select

**Comment by csandman** [comment]
There have been some major improvements in version 2 and version 3 of my [`chakra-react-select`](https://www.npmjs.com/package/chakra-react-select) package, allowing for a fully customizable style system, and the native typescript support added in `react-select@5`.  So if my component didn't feel customizable enough at first, or you had issues implementing it in a TypeScript project, I highly recommend giving it a second chance.

**Comment by rodrigoea** [comment]
>Here's a version I started building for an app using Chakra UI components, if anyone also wants it.
>
>https://github.com/chakra-ui/chakra-ui/assets/1905961/ead47f59-b5a6-4a64-bfbc-4d7559b9fa4d
>
>Here's the CodeSandbox for reference – it's just a simple component, and it might not be 100% optimized for your application yet 🤓 

>https://codesandbox.io/s/chakra-ui-multiselect-kn2gvz?file=/src/multi-select.js



from: https://github.com/chakra-ui/chakra-ui/issues/1293#issuecomment-1599296636
